sydent/db/sqlitedb.py

```diff
@@ -186,6 +186,15 @@ def _upgradeSchema(self):
             logger.info("v3 -> v4 schema migration complete")
             self._setSchemaVersion(4)
 
+        if curVer < 5:
+            # Fix lookup_hash index for selecting on mxid instead of medium
+            cur = self.db.cursor()
+            cur.execute("DROP INDEX lookup_hash_medium")
+            cur.execute("CREATE UNIQUE INDEX lookup_hash_mxid ON global_threepid_associations(lookup_hash, mxid)")
```

Comment:
It's just what you're selecting by that matters with the index, the multiple columns are the levels in the index tree.

 * For `SELECT FROM foo WHERE meep = 1` you'd want `CREATE INDEX floopity ON foo (meep)`
 * For `SELECT FROM foo WHERE meep = 1 and fweep = 2` you'd want `CREATE INDEX floopity ON foo (meep, fweep)`

Also even though we don't need this on the local table, we should still prefix with `global_threepid_` even if only for consistency with the others above.

Also, come to think of it, the `UNIQUE` part shouldn't be there either because apart from hashes not necessarily being 100% unique, this was initially designed as an append-only table and the lookup would just take the latest result (although now of course we remove the addresses to strip the PII - I'm unsure whether or not the same applies to the hashes).

So I think you want:

```suggestion
            cur.execute("CREATE INDEX global_threepid_lookup_hash ON global_threepid_associations(lookup_hash)")
```

Reply:
Great explanation, thank you!